README.md
## Swarmer — create, start and wait for Android emulators to boot.

`Swarmer` is a replacement for [such bash scripts](https://github.com/travis-ci/travis-cookbooks/blob/master/community-cookbooks/android-sdk/files/default/android-wait-for-emulator) but with features like:

* Start **multiple** Android Emulators and wait for each to boot in parallel.
* Pass `config.ini` that'll be applied to created emulator.
* Redirect Logcat output of an emulator to a file.

## How to use

`Swarmer` shipped as `jar`, to run it you need JVM 1.8+: `java -jar swarmer-latest-version.jar options`, you need to have Android SDK Tools 25.2.3+.

#### Supported options

##### Required

* `--emulator-name`
  * Name of the emulator, i.e. `test_emulator_1`.
* `--package`
  * Package of the system image for this AVD (e.g.'system-images;android-25;google_apis;x86') to pass to `avdmanager create avd --package`.
* `--android-abi`
  * Android system image abi, i.e. `google_apis/x86_64`.
* `--path-to-config-ini`
  * Path either relative or absolute to the file that will be used as `config.ini` for created emulator.

##### Optional

* `--help, -help, help, -h`
  * Print help and exit.
* `--emulator-start-options`
  * Options to pass to `emulator -avd \$emulatorName` command, i.e. `--no-window -prop persist.sys.language=en -prop persist.sys.country=US`.
* `--emulator-start-timeout-seconds`
  * Timeout to wait for emulator to finish boot. Default value is 180 seconds.
* `--redirect-logcat-to`
  * Path either relative or absolute to the file that will be used to redirect logcat of started emulator to. No redirection will happen if parameter is not presented.

##### Examples

###### Start one emulator

```console
java -jar swarmer-latest-version.jar \
--emulator-name test_emulator_1 \
--package "system-images;android-25;google_apis;x86" \
--android-abi google_apis/x86_64 \
--path-to-config-ini emulator_config.ini \
--emulator-start-options -prop persist.sys.language=en -prop persist.sys.country=US \
--redirect-logcat-to test_emulator_1_logcat.txt
```

###### Start two emulators in parallel

```console
java -jar swarmer-latest-version.jar \
--emulator-name test_emulator_1 \
--package "system-images;android-25;google_apis;x86" \
--android-abi google_apis/x86_64 \
--path-to-config-ini emulator_config1.ini \
--emulator-start-options -prop persist.sys.language=en -prop persist.sys.country=US \
--redirect-logcat-to test_emulator_1_logcat.txt
--emulator-name test_emulator_2 \
--package "system-images;android-23;google_apis;x86" \
--android-abi google_apis/x86_64 \
--path-to-config-ini emulator_config2.ini \
--emulator-start-options -prop persist.sys.language=en -prop persist.sys.country=US \
--redirect-logcat-to test_emulator_2_logcat.txt
```

###### Start two emulators sequentially

```console
java -jar swarmer-latest-version.jar \
--emulator-name test_emulator_1 \
--android-target android-25 \
--android-abi google_apis/x86_64 \
--path-to-config-ini emulator_config.ini \
--emulator-start-options -prop persist.sys.language=en -prop persist.sys.country=US \
--redirect-logcat-to test_emulator_1_logcat.txt

java -jar swarmer-latest-version.jar \
--emulator-name test_emulator_2 \
--package "system-images;android-23;google_apis;x86" \
--android-abi google_apis/x86_64 \
--path-to-config-ini emulator_config2.ini \
--emulator-start-options -prop persist.sys.language=en -prop persist.sys.country=US \
--redirect-logcat-to test_emulator_2_logcat.txt
```

### Download

Swarmer is [available on jcenter](https://jcenter.bintray.com/com/gojuno/swarmer).

>You can download it in your CI scripts or store it in your version control system (not recommended).

All the releases and changelogs can be found on [Releases Page](https://github.com/gojuno/swarmer/releases).

### How to build

Dependencies: you only need `docker` and `bash` installed on your machine.

```console
bash ci/build.sh
```

## License

```
Copyright 2017 Juno, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# See [Releases](https://github.com/gojuno/swarmer/releases)


swarmer/src/test/kotlin/com/gojuno/swarmer/ArgsSpec.kt
package com.gojuno.swarmer

import org.assertj.core.api.Assertions.assertThat
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on

class ArgsSpec : Spek({

    on("parse args with only required fields") {

        val result by memoized {
            parseArgs(listOf(
                    "--emulator-name", "test_emulator_name",
                    "--package", "test_android_package",
                    "--android-abi", "test_android_abi",
                    "--path-to-config-ini", "test_path_to_config_ini"
            ))
        }

        it("parses passed args and uses default values for non-required fields") {
            assertThat(result).isEqualTo(listOf(Args(
                    emulatorName = "test_emulator_name",
                    pakage = "test_android_package",
                    androidAbi = "test_android_abi",
                    pathToConfigIni = "test_path_to_config_ini",
                    emulatorStartOptions = emptyList(),
                    emulatorStartTimeoutSeconds = 180,
                    redirectLogcatTo = null
            )))
        }
    }

    on("parse multiple args") {

        val result by memoized {
            parseArgs(listOf(
                    "--emulator-name", "test_emulator_name_1",
                    "--package", "test_android_package_1",
                    "--android-abi", "test_android_abi_1",
                    "--path-to-config-ini", "test_path_to_config_ini_1",
                    "--emulator-name", "test_emulator_name_2",
                    "--package", "test_android_package_2",
                    "--android-abi", "test_android_abi_2",
                    "--path-to-config-ini", "test_path_to_config_ini_2"
            ))
        }

        it("parses two arguments") {
            assertThat(result).isEqualTo(listOf(
                    Args(
                            emulatorName = "test_emulator_name_1",
                            pakage = "test_android_package_1",
                            androidAbi = "test_android_abi_1",
                            pathToConfigIni = "test_path_to_config_ini_1",
                            emulatorStartOptions = emptyList(),
                            emulatorStartTimeoutSeconds = 180,
                            redirectLogcatTo = null
                    ),
                    Args(
                            emulatorName = "test_emulator_name_2",
                            pakage = "test_android_package_2",
                            androidAbi = "test_android_abi_2",
                            pathToConfigIni = "test_path_to_config_ini_2",
                            emulatorStartOptions = emptyList(),
                            emulatorStartTimeoutSeconds = 180,
                            redirectLogcatTo = null
                    )))
        }
    }
})


swarmer/src/test/kotlin/com/gojuno/janulator/ArgsSpec.kt
package com.gojuno.janulator

import org.assertj.core.api.Assertions.assertThat
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on

class ArgsSpec : Spek({

    on("parse args with only required fields") {

        val result by memoized {
            parseArgs(listOf(
                    "--emulator-name", "test_emulator_name",
                    "--package", "test_android_package",
                    "--android-abi", "test_android_abi",
                    "--path-to-config-ini", "test_path_to_config_ini"
            ))
        }

        it("parses passed args and uses default values for non-required fields") {
            assertThat(result).isEqualTo(listOf(Args(
                    emulatorName = "test_emulator_name",
                    pakage = "test_android_package",
                    androidAbi = "test_android_abi",
                    pathToConfigIni = "test_path_to_config_ini",
                    emulatorStartOptions = emptyList(),
                    emulatorStartTimeoutSeconds = 180,
                    redirectLogcatTo = null
            )))
        }
    }

    on("parse multiple args") {

        val result by memoized {
            parseArgs(listOf(
                    "--emulator-name", "test_emulator_name_1",
                    "--package", "test_android_package_1",
                    "--android-abi", "test_android_abi_1",
                    "--path-to-config-ini", "test_path_to_config_ini_1",
                    "--emulator-name", "test_emulator_name_2",
                    "--package", "test_android_package_2",
                    "--android-abi", "test_android_abi_2",
                    "--path-to-config-ini", "test_path_to_config_ini_2"
            ))
        }

        it("parses two arguments") {
            assertThat(result).isEqualTo(listOf(
                    Args(
                            emulatorName = "test_emulator_name_1",
                            pakage = "test_android_package_1",
                            androidAbi = "test_android_abi_1",
                            pathToConfigIni = "test_path_to_config_ini_1",
                            emulatorStartOptions = emptyList(),
                            emulatorStartTimeoutSeconds = 180,
                            redirectLogcatTo = null
                    ),
                    Args(
                            emulatorName = "test_emulator_name_2",
                            pakage = "test_android_package_2",
                            androidAbi = "test_android_abi_2",
                            pathToConfigIni = "test_path_to_config_ini_2",
                            emulatorStartOptions = emptyList(),
                            emulatorStartTimeoutSeconds = 180,
                            redirectLogcatTo = null
                    )))
        }
    }
})


ci/build.sh
#!/bin/bash
set -xe

# You can run it from any directory.
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
PROJECT_DIR="$DIR/.."

pushd "$PROJECT_DIR"

USER_ID=`id -u $USER`

BUILD_COMMAND="set -xe && "
BUILD_COMMAND+="apt-get update && apt-get -y install sudo && "
BUILD_COMMAND+="groupadd --gid $USER_ID build_user && "
BUILD_COMMAND+="useradd --shell /bin/bash --uid $USER_ID --gid $USER_ID --create-home build_user && "
BUILD_COMMAND+="sudo --set-home --preserve-env -u build_user "

BUILD_COMMAND+="/opt/project/gradlew "
BUILD_COMMAND+="--no-daemon --info "
BUILD_COMMAND+="clean build "

if [ "$PUBLISH" = "true" ]; then
    BUILD_COMMAND+="artifactoryPublish "
fi

BUILD_COMMAND+="--project-dir /opt/project"

docker run \
-v `"pwd"`:/opt/project \
openjdk:8u121-jdk \
bash -c "$BUILD_COMMAND"

popd


swarmer/src/main/kotlin/com/gojuno/swarmer/Main.kt
package com.gojuno.swarmer

import com.gojuno.commander.android.connectedAdbDevices
import com.gojuno.commander.os.log
import rx.Observable
import rx.schedulers.Schedulers.io
import java.lang.System.nanoTime
import java.util.concurrent.Semaphore

fun main(vararg rawArgs: String) {
    val args = parseArgs(rawArgs.toList())

    val startTime = nanoTime()

    // Sometimes on Linux "emulator -verbose -avd" does not print serial id of started emulator,
    // so by allocating ports manually we know which serial id emulator will have.
    val availablePortsSemaphore = Semaphore(1)

    val startedEmulators = connectedAdbDevices()
            .doOnNext { log("Already running emulators: $it") }
            .flatMap {
                val startEmulators: List<Observable<Emulator>> = args
                        .map { startEmulator(it, availablePortsSemaphore) }
                        .map { it.subscribeOn(io()) } // So each emulator will start in parallel.
                        .map { it.doOnNext { log("Emulator $it is ready.") } }

                Observable.zip(startEmulators, { startedEmulator -> startedEmulator })
            }
            .map { it.map { it as Emulator }.toSet() }
            .toBlocking()
            .firstOrDefault(emptySet())

    log("Swarmer: - \"My job is done here, took ${(nanoTime() - startTime).nanosAsSeconds()} seconds, startedEmulators: $startedEmulators, bye bye.\"")
    System.exit(0) // Force exit, emulator and logcat redirect will keep running as detached processes.
}


swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt
package com.gojuno.swarmer

import com.gojuno.commander.android.*
import com.gojuno.commander.os.Notification
import com.gojuno.commander.os.home
import com.gojuno.commander.os.log
import com.gojuno.commander.os.process
import rx.Observable
import rx.schedulers.Schedulers
import rx.schedulers.Schedulers.io
import java.io.File
import java.lang.System.nanoTime
import java.util.concurrent.Semaphore
import java.util.concurrent.TimeUnit.*
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicLong

val sh: String = "/bin/sh"
val avdManager: String = "$androidHome/tools/bin/avdmanager"
val emulator: String = "$androidHome/tools/emulator"

data class Emulator(
        val id: String,
        val name: String
)

fun startEmulator(args: Args, availablePortsSemaphore: Semaphore): Observable<Emulator> {
    return createAvd(args)
            .flatMap { applyConfig(args) }
            .map { availablePortsSemaphore.acquire() }
            .flatMap { findAvailablePortsForNewEmulator() }
            .doOnNext { log("Ports for emulator ${args.emulatorName}: ${it.first}, ${it.second}.") }
            .flatMap { ports ->
                val emulatorProcess = process(
                        // Unix only, PR welcome.
                        listOf(sh, "-c", "$emulator -verbose -avd ${args.emulatorName} -ports ${ports.first},${ports.second} ${args.emulatorStartOptions.joinToString(" ")} &"),
                        timeout = null,
                        redirectOutputTo = outputFileForEmulator(args)
                )
                waitForEmulatorToStart(args, emulatorProcess, ports)
            }
            .map { emulator -> availablePortsSemaphore.release().let { emulator } }
            .flatMap { emulator ->
                when (args.redirectLogcatTo) {
                    null -> Observable.just(emulator)
                    else -> {
                        val adbDevice = AdbDevice(id = emulator.id, online = true)
                        val logcatFile = File(args.redirectLogcatTo)

                        adbDevice
                                .redirectLogcatToFile(logcatFile)
                                .doOnSubscribe { adbDevice.log("Redirecting logcat output to file $logcatFile") }
                                .toObservable()
                                .subscribeOn(Schedulers.io())
                                .map { emulator }
                    }
                }
            }
            .flatMap { waitForEmulatorToFinishBoot(it) }
            .timeout(args.emulatorStartTimeoutSeconds, SECONDS)
            .doOnError {
                when (it) {
                    is TimeoutException -> println("Timeout ${args.emulatorStartTimeoutSeconds} seconds, failed to start emulator ${args.emulatorName}.")
                    else -> Unit
                }
            }
}

private fun createAvd(args: Args): Observable<Unit> {
    val createAvdProcess = process(
            listOf(
                    avdManager, "create",
                    "avd", "--force",
                    "--name", args.emulatorName,
                    "--package", args.pakage,
                    "--abi", args.androidAbi
            ),
            timeout = 60 to SECONDS
    ).share()

    val iDontWishToCreateCustomHardwareProfile = Observable
            .combineLatest(
                    createAvdProcess.filter { it is Notification.Start }.cast(Notification.Start::class.java),
                    Observable.interval(500, MILLISECONDS),
                    { notification, _ -> notification }
            )
            .first { it.output.readText().contains("Do you wish to create a custom hardware profile") }
            .observeOn(io())
            .map { it.process.outputStream.writer().use { it.write("no") } }

    val startTime = AtomicLong()

    return Observable
            .merge(iDontWishToCreateCustomHardwareProfile, createAvdProcess)
            .first { it is Notification.Exit }
            .doOnError { log("Error during creation of avd ${args.emulatorName}, error = $it") }
            .retry(3) // https://code.google.com/p/android/issues/detail?id=262719
            .map { Unit }
            .doOnSubscribe { log("Creating avd ${args.emulatorName}."); startTime.set(nanoTime()) }
            .doOnNext { log("Avd ${args.emulatorName} created in ${(nanoTime() - startTime.get()).nanosAsSeconds()} seconds.") }
            .doOnError { log("Could not create avd ${args.emulatorName}, error = $it") }
}

private fun applyConfig(args: Args): Observable<Unit> = Observable
        .fromCallable {
            File(args.pathToConfigIni)
                    .copyTo(File("$home/.android/avd/${args.emulatorName}.avd/config.ini"), overwrite = true)
        }
        .map { Unit }

private fun findAvailablePortsForNewEmulator(): Observable<Pair<Int, Int>> = connectedAdbDevices()
        .map { it.filter { it.isEmulator } }
        .map {
            if (it.isEmpty()) {
                5554
            } else {
                it
                        .map { it.id }
                        .map { it.substringAfter("emulator-") }
                        .map { it.toInt() }
                        .max()!!
                        .let { it + 2 }
            }
        }
        .map { it to it + 1 }

private fun waitForEmulatorToStart(args: Args, emulatorProcess: Observable<Notification>, ports: Pair<Int, Int>): Observable<Emulator> {
    val startTime = AtomicLong()

    return emulatorProcess
            .filter { it is Notification.Start }
            .cast(Notification.Start::class.java)
            .flatMap {
                // Wait for emulator serial number to show up in emulator -avd output.
                Observable
                        .interval(1000, MILLISECONDS)
                        .flatMap { connectedAdbDevices() }
                        .map {
                            if (it.map { it.id }.contains("emulator-${ports.first}")) {
                                Emulator(id = "emulator-${ports.first}", name = args.emulatorName)
                            } else {
                                null
                            }
                        }
                        .filter { it != null }
                        .first()
            }
            .cast(Emulator::class.java) // Lost Kotlin nullability after flatMap.
            .flatMap { emulator ->
                // Wait for emulator to show up in adb devices.
                Observable
                        .interval(1000, MILLISECONDS)
                        .flatMap { connectedAdbDevices() }
                        .map { it.firstOrNull { it.id == emulator.id } }
                        .filter { it != null }
                        .first()
                        .map { emulator }
            }
            .doOnSubscribe { log("Starting emulator ${args.emulatorName}."); startTime.set(nanoTime()) }
            .doOnNext { log("Emulator $it started in ${(nanoTime() - startTime.get()).nanosAsSeconds()} seconds.") }
            .doOnError { log("Error during start of the emulator ${args.emulatorName}, error = $it") }
}

private fun waitForEmulatorToFinishBoot(targetEmulator: Emulator): Observable<Emulator> {
    val startTime = AtomicLong()

    return Observable
            .interval(1500, MILLISECONDS)
            .switchMap { connectedAdbDevices() }
            .map { it.filter { it.isEmulator } }
            .switchMap { runningEmulators ->
                val emulator = runningEmulators.firstOrNull { it.id == targetEmulator.id }

                if (emulator == null || emulator.online == false) {
                    Observable.never()
                } else {
                    process(
                            listOf(
                                    adb,
                                    "-s", emulator.id,
                                    "shell",
                                    "getprop", "init.svc.bootanim"
                            ),
                            timeout = 10 to SECONDS
                    )
                            .filter { it is Notification.Exit }
                            .cast(Notification.Exit::class.java)
                            .map { it.output.readText().contains("stopped", ignoreCase = true) }
                            .switchMap { bootAnimationStopped ->
                                if (bootAnimationStopped) {
                                    Observable.just(targetEmulator)
                                } else {
                                    Observable.never()
                                }
                            }
                }
            }
            .first()
            .doOnSubscribe { log("Waiting boot process to finish for emulator $targetEmulator."); startTime.set(nanoTime()) }
            .doOnNext { log("Emulator $targetEmulator finished boot process in ${(nanoTime() - startTime.get()).nanosAsSeconds()} seconds.") }
            .doOnError { log("Error during start of the emulator $targetEmulator, error = $it.") }
}

fun Long.nanosAsSeconds(): Float = NANOSECONDS.toMillis(this) / 1000f

private fun outputFileForEmulator(args: Args) = File("${args.emulatorName}.output")
        .apply { deleteOnExit() }


swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt
package com.gojuno.swarmer

import com.beust.jcommander.JCommander
import com.beust.jcommander.Parameter

data class Args(
        val emulatorName: String,
        val pakage: String,
        val androidAbi: String,
        val pathToConfigIni: String,
        val emulatorStartOptions: List<String>,
        val emulatorStartTimeoutSeconds: Long,
        val redirectLogcatTo: String?
)

// No way to share array both for runtime and annotation without reflection.
private val PARAMETER_HELP_NAMES = setOf("--help", "-help", "help", "-h")
private const val PARAMETER_EMULATOR_NAME = "--emulator-name"

private class JCommanderArgs {

    @Parameter(
            names = arrayOf("--help", "-help", "help", "-h"),
            help = true,
            description = "Print help and exit."
    )
    var help: Boolean? = null

    @Parameter(
            names = arrayOf(PARAMETER_EMULATOR_NAME),
            required = true,
            description = "Name for the emulator, i.e. `test_emulator_1` to pass to `avdmanager create avd --name`."
    )
    lateinit var emulatorName: String

    @Parameter(
            names = arrayOf("--package"),
            required = true,
            description = "Package of the system image for this AVD (e.g.'system-images;android-25;google_apis;x86') to pass to `avdmanager create avd --package`."
    )
    lateinit var pakage: String

    @Parameter(
            names = arrayOf("--android-abi"),
            required = true,
            description = "Android system image abi, i.e. `google_apis/x86_64` to pass to `avdmanager create avd --abi`."
    )
    lateinit var androidAbi: String

    @Parameter(
            names = arrayOf("--path-to-config-ini"),
            required = true,
            description = "Path either relative or absolute to the file that will be used as `config.ini` for created emulator."
    )
    lateinit var pathToConfigIni: String

    @Parameter(
            names = arrayOf("--emulator-start-options"),
            required = false,
            variableArity = true,
            description = "Options to pass to `emulator -avd \$emulatorName` command, i.e. `--no-window -prop persist.sys.language=en -prop persist.sys.country=US`."
    )
    var emulatorStartOptions: List<String>? = null

    @Parameter(
            names = arrayOf("--emulator-start-timeout-seconds"),
            required = false,
            description = "Timeout to wait for emulator to finish boot. Default value is 180 seconds."
    )
    var emulatorStartTimeoutSeconds: Long? = null

    @Parameter(
            names = arrayOf("--redirect-logcat-to"),
            required = false,
            description = "Path either relative or absolute to the file that will be used to redirect logcat of started emulator to. No redirection will happen if parameter is not presented.")
    var redirectLogcatTo: String? = null
}

fun parseArgs(rawArgs: List<String>): List<Args> {
    if (PARAMETER_HELP_NAMES.firstOrNull { rawArgs.contains(it) } != null) {
        JCommander(JCommanderArgs()).usage()
        System.exit(0)
    }

    // PR to rewrite this mess to Rx or Kotlin Stream API very welcome.

    fun List<String>.indexOfFirstParameterEmulatorName(): Int = indexOfFirst { it == PARAMETER_EMULATOR_NAME }

    fun List<String>.indexOfSecondParameterEmulatorName(): Int {
        val indexOfFirst = indexOfFirstParameterEmulatorName()

        return when (indexOfFirst) {
            -1 -> -1
            else -> {
                if (indexOfFirst == 0 && size > 1) {
                    subList(indexOfFirst + 1, size).indexOfFirstParameterEmulatorName().let { if (it == -1) it else it + 1 }
                } else {
                    -1
                }
            }
        }
    }

    var indexOfFirstParameterEmulatorName = rawArgs.indexOfFirstParameterEmulatorName()
    var indexOfSecondParameterEmulatorName = rawArgs.indexOfSecondParameterEmulatorName()

    var argsLeft = rawArgs
    var splitArgs = emptyList<List<String>>()

    while (indexOfFirstParameterEmulatorName != -1) {
        splitArgs += listOf(argsLeft.subList(indexOfFirstParameterEmulatorName, indexOfSecondParameterEmulatorName.let { if (it == -1) argsLeft.size else it }))
        argsLeft = if (indexOfSecondParameterEmulatorName == -1)
            emptyList()
        else
            argsLeft.subList(indexOfSecondParameterEmulatorName, argsLeft.size)

        indexOfFirstParameterEmulatorName = argsLeft.indexOfFirstParameterEmulatorName()
        indexOfSecondParameterEmulatorName = argsLeft.indexOfSecondParameterEmulatorName()
    }

    return splitArgs.map { parseEachArgSet(it) }
}

private fun parseEachArgSet(rawArgs: List<String>): Args {
    val jCommanderArgs = JCommanderArgs()

    JCommander(jCommanderArgs, *rawArgs.toTypedArray())

    return jCommanderArgs.let {
        Args(
                emulatorName = it.emulatorName,
                pakage = it.pakage,
                androidAbi = it.androidAbi,
                pathToConfigIni = it.pathToConfigIni,
                emulatorStartOptions = it.emulatorStartOptions ?: emptyList(),
                emulatorStartTimeoutSeconds = it.emulatorStartTimeoutSeconds ?: 180,
                redirectLogcatTo = it.redirectLogcatTo
        )
    }
}

